File: src/main/java/com/springboot/couchbase/springbootrealworld/domain/article/repository/FavoriteRepository.java
package com.springboot.couchbase.springbootrealworld.domain.article.repository;


import com.springboot.couchbase.springbootrealworld.domain.article.entity.FavoriteDocument;
import org.springframework.data.couchbase.repository.Collection;
import org.springframework.data.couchbase.repository.Query;
import org.springframework.data.repository.CrudRepository;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;

@Repository
@Collection(FavoriteDocument.FAVORITE_COLLECTION_NAME)
public interface FavoriteRepository extends CrudRepository<FavoriteDocument, String> {
    @Query("#{#n1ql.selectEntity} WHERE id IS NOT NULL AND #{#n1ql.filter}")
    List<FavoriteDocument> findBySlug(String slug);

    Optional<FavoriteDocument> findByArticleIdAndAuthorEmail(String articleId, String author);


    Optional<FavoriteDocument> findByAuthorEmail(String author);

    List<FavoriteDocument> findByArticleId(String articleId);

    @Query("#{#n1ql.selectEntity} WHERE author IS NOT NULL AND #{#n1ql.filter}")
    List<FavoriteDocument> findAllFavorites();


}

File: src/main/java/com/springboot/couchbase/springbootrealworld/domain/article/repository/FavoriteRepository.kt
package com.springboot.couchbase.springbootrealworld.domain.article.repository

import com.springboot.couchbase.springbootrealworld.domain.article.entity.FavoriteDocument
import org.springframework.data.couchbase.repository.Collection
import org.springframework.data.couchbase.repository.Query;
import org.springframework.data.repository.CrudRepository
import org.springframework.stereotype.Repository

@Repository
@Collection(FavoriteDocument.FAVORITE_COLLECTION_NAME)
interface FavoriteRepository : CrudRepository<FavoriteDocument, String> {

    @Query("#{#n1ql.selectEntity} WHERE id IS NOT NULL AND #{#n1ql.filter}")
    fun findBySlug(slug: String): List<FavoriteDocument>

    fun findByArticleIdAndAuthorEmail(articleId: String, author: String): FavoriteDocument?

    fun findByAuthorEmail(author: String): FavoriteDocument?

    fun findByArticleId(articleId: String): List<FavoriteDocument>

    @Query("#{#n1ql.selectEntity} WHERE author IS NOT NULL AND #{#n1ql.filter}")
    fun findAllFavorites(): List<FavoriteDocument>
}
